### PR TITLE
Add `hasParameter` and `hasNoParameter` assertion methods for URIs and URLs

### PR DESCRIPTION
**Pull request by benblank**
This allows assertions in the style of:
 
    assertThat(new URI("http://foo.com/?bar=baz")).hasParameter("bar");

Additionally, support is available for a few edge cases I've seen in the wild; specifically, parameters with no value (e.g. `?foo&bar`) as well as multiple parameters with the same name.  Therefore, it can be meaningful to write an assertion like:

    assertThat(new URI("http://foo.com/?bar&bar=baz")).hasParameter("bar", null).hasParameter("bar", "baz");

#### Check List:
* Fixes : NA
* Unit tests : YES
* Javadoc with a code example (API only) : YES

### CODE REVIEW

**Comment by joel-costigliola** [comment]
Good idea, thanks for the PR. WIll review it soon

File: src/main/java/org/assertj/core/api/AbstractUrlAssert.java [hunk 1]
```diff
@@ -291,4 +291,104 @@ public S hasNoUserInfo() {
...
+  }
+
+  /**
+   * Verifies that the actual {@code URL} does not have a parameter with the specified name.
+   * <p>
+   * The value of the parameter is not checked.
+   * <p>
+   * Examples:
+   * <pre><code class='java'> // This assertion succeeds:
+   * assertThat(new URL("http://www.helloworld.org/index.html")).hasParameter("happy");
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractUrlAssert.java]: this is a valid a code example for `hasNoParameter` except that it shows `hasParameter` :p

File: src/main/java/org/assertj/core/api/AbstractUrlAssert.java [hunk 2]
```diff
@@ -291,4 +291,104 @@ public S hasNoUserInfo() {
...
+   * <p>
+   * Use {@code null} to indicate an absent value (e.g. {@code foo&bar}) as opposed to an empty value (e.g.
+   * {@code foo=&bar=}).
+   * <p>
+   * Examples:
+   * <pre><code class='java'> // These assertions succeed:
+   * assertThat(new URL("http://www.helloworld.org/index.html")).hasParameter("happy", "very");
+   * assertThat(new URL("http://www.helloworld.org/index.html?happy")).hasParameter("happy", "very");
+   * assertThat(new URL("http://www.helloworld.org/index.html?happy=very")).hasParameter("happy", null);
+   *
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractUrlAssert.java]: same comment no `hasNoParameter` used here

File: src/main/java/org/assertj/core/api/AbstractUriAssert.java [hunk 3]
```diff
@@ -287,4 +287,104 @@ public S hasNoUserInfo() {
...
+    return myself;
+  }
+
+  /**
+   * Verifies that the actual {@code URI} does not have a parameter with the expected name and value.
+   * <p>
+   * Use {@code null} to indicate an absent value (e.g. {@code foo&bar}) as opposed to an empty value (e.g.
+   * {@code foo=&bar=}).
+   * <p>
+   * Examples:
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractUriAssert.java]: use `hasNoParameter` in the examples

File: src/main/java/org/assertj/core/api/AbstractUriAssert.java [hunk 4]
```diff
@@ -287,4 +287,104 @@ public S hasNoUserInfo() {
...
+    return myself;
+  }
+
+  /**
+   * Verifies that the actual {@code URI} does not have a parameter with the specified name.
+   * <p>
+   * The value of the parameter is not checked.
+   * <p>
+   * Examples:
+   * <pre><code class='java'> // This assertion succeeds:
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractUriAssert.java]: use `hasNoParameter` in the examples

File: src/main/java/org/assertj/core/error/uri/ShouldHaveParameter.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+public class ShouldHaveParameter extends BasicErrorMessageFactory {
+
+  private static final String SHOULD_HAVE_PARAMETER_BUT_MISSING = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nbut was absent";
+  private static final String SHOULD_HAVE_PARAMETER_NO_VALUE_BUT_MISSING = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith no value, but was absent";
+  private static final String SHOULD_HAVE_PARAMETER_NO_VALUE_BUT_VALUE = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith no value, but had value:%n  <%s>";
+  private static final String SHOULD_HAVE_PARAMETER_VALUE_BUT_MISSING = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith value:%n  <%s>%nbut was absent";
+  private static final String SHOULD_HAVE_PARAMETER_VALUE_BUT_NO_VALUE = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith value:%n  <%s>%nbut had no value";
+  private static final String SHOULD_HAVE_PARAMETER_VALUE_BUT_WRONG = "%nExpecting:%n  <%s>%nto have parameter:%n  <%s>%nwith value:%n  <%s>%nbut had value:%n  <%s>";
+  private static final String SHOULD_HAVE_NO_PARAMETER_BUT_NO_VALUE = "%nExpecting:%n  <%s>%nto not have parameter:%n  <%s>%nbut was present with no value";
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/error/uri/ShouldHaveParameter.java]: I feel that `not to have a parameter` is better than `to not have parameter`

File: src/main/java/org/assertj/core/internal/Uris.java [hunk 6]
```diff
@@ -81,4 +89,91 @@ public void assertHasUserInfo(AssertionInfo info, URI actual, String expected) {
...
+    Map<String, List<String>> parameters = getParameters(actual.getQuery());
+
+    if (!parameters.containsKey(name)) {
+      throw failures.failure(info, shouldHaveParameter(actual, name, value));
+    }
+
+    List<String> values = parameters.get(name);
+
+    if (!values.contains(value)) {
+      throw failures.failure(info, shouldHaveParameter(actual, name, value, values.get(0)));
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Uris.java]: are we sure values can't be empty ?
If we display parameter's values, we should display them all between `[]`, if there is only one, stick with the actual message.

File: src/main/java/org/assertj/core/internal/Uris.java [hunk 7]
```diff
@@ -81,4 +89,91 @@ public void assertHasUserInfo(AssertionInfo info, URI actual, String expected) {
...
+
+  public void assertHasNoParameter(AssertionInfo info, URI actual, String name) {
+    assertNotNull(info, actual);
+
+    Map<String, List<String>> parameters = getParameters(actual.getQuery());
+
+    if (parameters.containsKey(name)) {
+      List<String> values = parameters.get(name);
+
+      throw failures.failure(info, shouldHaveNoParameter(actual, name, values.get(0)));
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/internal/Uris.java]: same comment here on displaying multiple values.

**Comment by benblank** [comment]
* **Bad examples** — Well, that's embarrassing!  Fixed.
* **Message phrasing** — No objection.  Changed.
* **Multiple values** — Good idea.  For single values, the messages now show it verbatim, but multiple values are shown "array style".  I've added a few tests to cover these cases, as well.

File: src/main/java/org/assertj/core/api/AbstractUriAssert.java [hunk 8]
```diff
@@ -287,4 +287,104 @@ public S hasNoUserInfo() {
...
+   * assertThat(new URI("http://www.helloworld.org/index.html")).hasParameter("happy");
```
Comment (PascalSchumacher) [inline src/main/java/org/assertj/core/api/AbstractUriAssert.java]: @benblank I guess you missed this line when fixing the javadoc examples?

**Comment by joel-costigliola** [comment]
Thanks a lot @benblank, I have integrated your work with (I believe) few improvements and an additional assertion : `hasNoParameters()`.

I'm closing this PR but feel free to comment my changes if you feel they can be improved.

**Pull request closed without merging**